docs/src/description/constraints.md

```diff
@@ -2,8 +2,158 @@
+
+To use public inputs, the public input must be declared in the `public_inputs` source section. They can be accessed using array indexing syntax, as described by the [accessor syntax rules](./syntax.md#section-specific-accessors).
+
+Random values can be accessed by using array indexing syntax on the `$rand` built-in, as described by the [accessor syntax rules](./syntax.md#section-specific-accessors).
+
+### Example of boundary constraints with public inputs and random values
+
+The following is an example of a valid `boundary_constraints` block that uses public inputs and random values:
+
+```
+def BoundaryConstraintsExample
+
+trace_columns:
+    main: [a, b]
+    aux: [p0, p1]
+
+public_inputs:
+    stack_inputs: [16]
+    stack_outputs: [16]
+
+boundary_constraints:
+    # these are main constraints that use public input values.
+    enf a.first = stack_inputs[0]
+    enf a.last = stack_outputs[0]
+
+    # these are auxiliary constraints that use public input values.
+    enf p0.first = stack_inputs[1]
+    enf p0.last = stack_outputs[1]
+
+    # these are auxiliary constraints that use random values from the verifier.
+    enf b.first = a + $rand[0]
+    enf p1.first = (stack_inputs[2] + $rand[0]) * (stack_inputs[3] + $rand[1])
+```
 
 ## Transition constraints (`transition_constraints`)
 
-TODO
+The `transition_constraints` section consists of expressions describing constraints that must be true at each row of the execution trace in order for the proof to be valid.
+
+**Transition constraints are optional.** It is equally valid to define an empty `transition_constraints` section or to omit the `transition_constraints` section entirely.
+
+Transition constraints that are defined against auxiliary columns or that use random values from the built-in `$rand` array will be identified as auxiliary constraints.
+
+A transition constraint definition must:
+
+1. start with a block indentation and the `enf` keyword to indicate that the constraint must be _enforced_.
+2. continue with an equality expression that describes the constraint. The expression may include numbers, trace columns, periodic columns, random values, and any of the available [operations](./syntax.md#operations).
+3. end with a newline.
+
+### Current and next rows
+
+Transition constraints have access to values in the "current" row of the trace to which the constraint is being applied, as well as the "next" row of the trace. The value of a trace column in the next row is specified with the `'` postfix operator, as described by the [accessor syntax rules](./syntax.md#section-specific-accessors).
+
+### Simple example of transition constraints
+
+The following is a simple example of a valid `transition_constraints` block using values from the current and next rows of the main and auxiliary traces:
+
+```
+def TransitionConstraintsExample
+
+trace_columns:
+    main: [a, b]
+    aux: [p]
+
+public_inputs:
+    # at least one public inputs is required.
+    program_hash: [4]
```

Comment:
nit: since we don't care about what is actually here should we say something like:
```
public_inputs:
    <omitted for bravety>
```